lib/src/path.cc

```diff
@@ -0,0 +1,219 @@
+#include <internal/path.hpp>
+#include <internal/config_exception.hpp>
+#include <internal/path_builder.hpp>
+#include <internal/config_util.hpp>
+#include <internal/path_parser.hpp>
+
+using namespace std;
+
+namespace hocon {
+
+    path::path() { }
+
+    path::path(List<shared_string> other_path) : _path(move(other_path)) { }
+
+    path::path(string first, path const& remainder) :
+            _path(make_shared<string>(move(first)), remainder._path) { }
+
+    path::path(vector<string> elements) {
+        if (elements.size() == 0) {
+            throw config_exception("Empty path");
+        }
+
+        path remainder;
+        if (elements.size() > 1) {
+            path_builder builder;
+            for (size_t i = 1; i < elements.size(); i++) {
+                builder.append_key(elements[i]);
+            }
+            remainder = builder.result();
+        }
+        _path = List<shared_string>(make_shared<string>(move(elements[0])), remainder._path);
+    }
+
+    path::path(vector<path> paths_to_concat) {
+        if (paths_to_concat.size() == 0) {
+            throw config_exception("Empty path");
+        }
+
+        path_builder builder;
+        if (paths_to_concat[0].has_remainder()) {
+            builder.append_path(paths_to_concat[0].remainder());
+        }
+
+        for (size_t i = 1; i < paths_to_concat.size(); i++) {
```

Comment:
Can you start at some other element besides the first one? This loop needs to start at index 1.